File: ml_pipeline/polymorph_lamr/bench/spandecode.py
"""Span-aware (chunk-level) decode for the LaMR pruner.

Token-level global top-k decode has a *conjunction failure* on multi-word
needles: it drops a span the moment ANY single token in it crosses the drop
threshold. A free-text root-cause phrase like ``"disk controller firmware
deadlock"`` survives only if every one of its tokens survives, so a single
confidently-droppable filler token ("the", a space) fragments the phrase and the
needle is lost. This is the ChunkKV (ICLR 2025) observation: "existing methods
overlook semantic relationships between tokens resulting in fragmented context."

The fix is to make keep/drop decisions at the granularity of semantic SPANS
(whitespace-delimited words by default, or fixed N-token chunks), dropping whole
spans most-droppable-first until the token budget is hit, never splitting a span.

The governing objective is HIGH RECALL — a false drop of a needle costs far more
than a false keep. The intuitive guess is that ``min`` aggregation (drop a span
only if EVERY token is confidently droppable) maximizes recall by letting one
salient token protect a whole phrase. **It does not.** Measured on the
answer-survival benchmark (R=0.5), the aggregators rank ``max`` 68.5% > ``mean``
62.2% >> ``min`` 23.9%. With a CALIBRATED per-token model, ``min`` is pathological:
it refuses to drop almost every span (most spans contain at least one low-prob
token), so to still hit the target rate it is forced to drop whichever spans
happen to be uniformly mid/high prob — which include real needles — while
keeping confident filler. ``max`` instead trusts the model: it drops the spans
the model flags as droppable and leaves the rest, which preserves more needles at
the same budget. So the default per-span aggregator is ``max`` (matching the Rust
runtime's Word+Max default); ``min`` is retained only as a deliberately
conservative option and for the unit tests that exercise its semantics.
"""

from __future__ import annotations

from functools import lru_cache

import tiktoken

from ..label.align import decode_tokens, encode_with_spans


@lru_cache(maxsize=1)
def _enc() -> tiktoken.Encoding:
    return tiktoken.get_encoding("cl100k_base")


# Whitespace bytes. cl100k attaches a leading space to the *following* token, so
# a word boundary is detected by a non-space->... transition where the token's
# first byte is whitespace (it begins a new word) or the previous token ended in
# a hard whitespace break.
_WS_BYTES = frozenset(b" \t\r\n")


def _word_spans(text: str, spans: list[tuple[int, int]]) -> list[list[int]]:
    """Group token indices into whitespace-delimited words.

    A new word begins whenever a token starts on (or follows) a whitespace
    boundary. cl100k folds a leading space into the next token, so the rule is:
    a token begins a new word if its first byte is whitespace, OR the previous
    token ended in a whitespace byte (a hard newline/tab break), OR it is the
    first token. Whitespace-only tokens attach to the word they precede.
    """
    raw = text.encode("utf-8")
    groups: list[list[int]] = []
    prev_end_ws = True  # before the first token, treat as a boundary
    for i, (s, e) in enumerate(spans):
        if e <= s:
            # empty span — fold into current group if any, else start one
            if groups:
                groups[-1].append(i)
            else:
                groups.append([i])
            continue
        first_ws = raw[s] in _WS_BYTES
        starts_word = (i == 0) or first_ws or prev_end_ws
        if starts_word or not groups:
            groups.append([i])
        else:
            groups[-1].append(i)
        prev_end_ws = raw[e - 1] in _WS_BYTES
    return groups


def _chunk_spans(n_tokens: int, chunk_size: int) -> list[list[int]]:
    """Fixed runs of ``chunk_size`` consecutive token indices (the ChunkKV unit)."""
    chunk_size = max(1, chunk_size)
    return [list(range(i, min(i + chunk_size, n_tokens))) for i in range(0, n_tokens, chunk_size)]


def _parse_granularity(span: str, text: str, spans: list[tuple[int, int]], n_tokens: int) -> list[list[int]]:
    if span == "word":
        return _word_spans(text, spans)
    if span.startswith("chunk:"):
        try:
            n = int(span.split(":", 1)[1])
        except ValueError as exc:  # pragma: no cover - guarded by callers
            raise ValueError(f"bad chunk granularity {span!r}") from exc
        return _chunk_spans(n_tokens, n)
    if span == "token":
        return [[i] for i in range(n_tokens)]
    raise ValueError(f"unknown span granularity {span!r} (use 'word' or 'chunk:N')")


_AGGREGATORS = {
    "min": min,   # conservative: drop a span only if EVERY token is droppable
    "mean": lambda xs: sum(xs) / len(xs),
    "max": max,   # default: drop a span if ANY token is droppable — wins on survival
}


def span_decode(
    ids: list[int],
    spans: list[tuple[int, int]],
    text: str,
    drop_probs,
    target_drop_rate: float,
    span: str = "word",
    aggregator: str = "max",
    force_keep: list[bool] | None = None,
    tokenizer: str = "cl100k",
) -> str:
    """Span-aware decode: group tokens into spans, drop whole spans (most-droppable
    first by aggregated drop-prob) until ~``target_drop_rate`` of tokens are gone,
    never splitting a span and never dropping a span that contains a force-kept
    token.

    Args:
      ids:           cl100k token ids.
      spans:         (start_byte, end_byte) per token (from ``encode_with_spans``).
      text:          the original text (for word-boundary detection).
      drop_probs:    per-token P(drop) (any sequence indexable by position).
      target_drop_rate: fraction of TOKENS to remove.
      span:          granularity — 'word' (default) or 'chunk:N' or 'token'.
      aggregator:    per-span score from its token drop-probs — 'max' (default,
                     best survival), 'mean', or 'min' (conservative).
      force_keep:    optional per-token mask; any span containing a True is locked.

    Returns the decoded text of the surviving tokens.
    """
    n = len(ids)
    if n == 0:
        return decode_tokens(ids, tokenizer)
    if aggregator not in _AGGREGATORS:
        raise ValueError(f"unknown aggregator {aggregator!r} (use 'min'/'mean'/'max')")
    agg = _AGGREGATORS[aggregator]
    fk = force_keep or [False] * n

    groups = _parse_granularity(span, text, spans, n)

    # Budget: how many tokens we want to drop.
    k = max(0, round(max(0.0, min(1.0, target_drop_rate)) * n))

    # Score each span; locked spans are never droppable.
    scored: list[tuple[float, int, list[int]]] = []
    for gi, g in enumerate(groups):
        if any(fk[i] for i in g):
            continue  # force-kept span — never dropped
        score = agg([float(drop_probs[i]) for i in g])
        scored.append((score, gi, g))

    # Most-droppable first: highest aggregated drop-prob. Tie-break by span order
    # for determinism.
    scored.sort(key=lambda t: (-t[0], t[1]))

    dropped: set[int] = set()
    for _score, _gi, g in scored:
        if len(dropped) >= k:
            break
        # Atomic: only drop the whole span if it fits in the remaining budget,
        # otherwise skip it and try a smaller span (preserves the rate without
        # splitting a phrase). This keeps achieved-drop close to target while
        # honoring the no-split invariant.
        if len(dropped) + len(g) > k:
            continue
        dropped.update(g)

    return decode_tokens(
        [tid for i, tid in enumerate(ids) if i not in dropped], tokenizer
    )


def span_decode_from_text(
    text: str,
    drop_probs,
    target_drop_rate: float,
    span: str = "word",
    aggregator: str = "max",
    force_keep: list[bool] | None = None,
    tokenizer: str = "cl100k",
) -> str:
    """Convenience wrapper: derive ids/spans from ``text`` via ``encode_with_spans``
    then ``span_decode``. ``drop_probs`` must be aligned to the encoded ids.
    """
    ids, spans = encode_with_spans(text, tokenizer)
    return span_decode(
        ids, spans, text, drop_probs, target_drop_rate,
        span=span, aggregator=aggregator, force_keep=force_keep, tokenizer=tokenizer,
    )

File: src/spandecode.rs
//! Span-aware (chunk-level) decode for the LaMR pruner.
//!
//! Token-level global top-k decode has a *conjunction failure* on multi-word
//! needles: it drops a span the moment ANY single token in it crosses the drop
//! threshold, fragmenting multi-word phrases. The fix is to make keep/drop
//! decisions at the granularity of semantic SPANS (whitespace-delimited words by
//! default, or fixed N-token chunks), dropping whole spans most-droppable-first
//! until the token budget is hit, never splitting a span.
//!
//! The default per-span aggregator is `max` (matching the Rust runtime's
//! Word+Max default): drop a span if ANY token is droppable — empirically best
//! on the answer-survival benchmark (max 68.5% > mean 62.2% >> min 23.9% at
//! R=0.5). Ported from the Python `polymorph_lamr.bench.spandecode`.

use anyhow::{anyhow, Result};

use crate::tokenizer::{decode_tokens, token_spans};

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum Aggregator {
    Min,
    Mean,
    Max,
}

impl Aggregator {
    pub fn parse(s: &str) -> Result<Aggregator> {
        match s {
            "min" => Ok(Aggregator::Min),
            "mean" => Ok(Aggregator::Mean),
            "max" => Ok(Aggregator::Max),
            other => Err(anyhow!(
                "unknown aggregator {other:?} (use 'min'/'mean'/'max')"
            )),
        }
    }

    fn apply(self, xs: &[f64]) -> f64 {
        match self {
            Aggregator::Min => xs.iter().cloned().fold(f64::INFINITY, f64::min),
            Aggregator::Max => xs.iter().cloned().fold(f64::NEG_INFINITY, f64::max),
            Aggregator::Mean => xs.iter().sum::<f64>() / xs.len() as f64,
        }
    }
}

const WS_BYTES: [u8; 4] = [b' ', b'\t', b'\r', b'\n'];

fn is_ws(b: u8) -> bool {
    WS_BYTES.contains(&b)
}

/// Group token indices into whitespace-delimited words. cl100k folds a leading
/// space into the next token, so a token begins a new word if its first byte is
/// whitespace, OR the previous token ended in a whitespace byte, OR it is the
/// first token.
pub fn word_spans(text: &str, spans: &[(usize, usize)]) -> Vec<Vec<usize>> {
    let raw = text.as_bytes();
    let mut groups: Vec<Vec<usize>> = Vec::new();
    let mut prev_end_ws = true;
    for (i, &(s, e)) in spans.iter().enumerate() {
        if e <= s {
            if let Some(last) = groups.last_mut() {
                last.push(i);
            } else {
                groups.push(vec![i]);
            }
            continue;
        }
        let first_ws = is_ws(raw[s]);
        let starts_word = i == 0 || first_ws || prev_end_ws;
        if starts_word || groups.is_empty() {
            groups.push(vec![i]);
        } else {
            groups.last_mut().unwrap().push(i);
        }
        prev_end_ws = is_ws(raw[e - 1]);
    }
    groups
}

/// Fixed runs of `chunk_size` consecutive token indices (the ChunkKV unit).
pub fn chunk_spans(n_tokens: usize, chunk_size: usize) -> Vec<Vec<usize>> {
    let chunk_size = chunk_size.max(1);
    let mut out = Vec::new();
    let mut i = 0;
    while i < n_tokens {
        out.push((i..(i + chunk_size).min(n_tokens)).collect());
        i += chunk_size;
    }
    out
}

fn parse_granularity(
    span: &str,
    text: &str,
    spans: &[(usize, usize)],
    n_tokens: usize,
) -> Result<Vec<Vec<usize>>> {
    if span == "word" {
        return Ok(word_spans(text, spans));
    }
    if let Some(rest) = span.strip_prefix("chunk:") {
        let n: usize = rest
            .parse()
            .map_err(|_| anyhow!("bad chunk granularity {span:?}"))?;
        return Ok(chunk_spans(n_tokens, n));
    }
    if span == "token" {
        return Ok((0..n_tokens).map(|i| vec![i]).collect());
    }
    Err(anyhow!(
        "unknown span granularity {span:?} (use 'word' or 'chunk:N')"
    ))
}

/// Span-aware decode: group tokens into spans, drop whole spans (most-droppable
/// first by aggregated drop-prob) until ~`target_drop_rate` of tokens are gone,
/// never splitting a span and never dropping a span containing a force-kept token.
#[allow(clippy::too_many_arguments)]
pub fn span_decode(
    ids: &[u32],
    spans: &[(usize, usize)],
    text: &str,
    drop_probs: &[f64],
    target_drop_rate: f64,
    span: &str,
    aggregator: Aggregator,
    force_keep: Option<&[bool]>,
) -> Result<String> {
    let n = ids.len();
    if n == 0 {
        return decode_tokens(ids);
    }
    let fk_owned;
    let fk: &[bool] = match force_keep {
        Some(f) => f,
        None => {
            fk_owned = vec![false; n];
            &fk_owned
        }
    };

    let groups = parse_granularity(span, text, spans, n)?;

    let rate = target_drop_rate.clamp(0.0, 1.0);
    let k = (rate * n as f64).round() as usize;

    // Score each span; locked spans are never droppable.
    let mut scored: Vec<(f64, usize, &Vec<usize>)> = Vec::new();
    for (gi, g) in groups.iter().enumerate() {
        if g.iter().any(|&i| fk[i]) {
            continue;
        }
        let vals: Vec<f64> = g.iter().map(|&i| drop_probs[i]).collect();
        scored.push((aggregator.apply(&vals), gi, g));
    }

    // Most-droppable first: highest aggregated drop-prob, tie-break by span order.
    scored.sort_by(|a, b| {
        b.0.partial_cmp(&a.0)
            .unwrap_or(std::cmp::Ordering::Equal)
            .then(a.1.cmp(&b.1))
    });

    let mut dropped: std::collections::HashSet<usize> = std::collections::HashSet::new();
    for (_score, _gi, g) in &scored {
        if dropped.len() >= k {
            break;
        }
        if dropped.len() + g.len() > k {
            continue;
        }
        for &i in g.iter() {
            dropped.insert(i);
        }
    }

    let survivors: Vec<u32> = ids
        .iter()
        .enumerate()
        .filter(|(i, _)| !dropped.contains(i))
        .map(|(_, &tid)| tid)
        .collect();
    decode_tokens(&survivors)
}

/// Convenience wrapper: derive ids/spans from `text` then `span_decode`.
pub fn span_decode_from_text(
    text: &str,
    drop_probs: &[f64],
    target_drop_rate: f64,
    span: &str,
    aggregator: Aggregator,
    force_keep: Option<&[bool]>,
) -> Result<String> {
    let (ids, spans) = token_spans(text)?;
    span_decode(
        &ids,
        &spans,
        text,
        drop_probs,
        target_drop_rate,
        span,
        aggregator,
        force_keep,
    )
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn chunk_spans_partition() {
        let g = chunk_spans(7, 3);
        assert_eq!(g, vec![vec![0, 1, 2], vec![3, 4, 5], vec![6]]);
    }

    #[test]
    fn word_spans_group_by_whitespace() {
        let text = "alpha beta gamma";
        let (_ids, spans) = token_spans(text).unwrap();
        let groups = word_spans(text, &spans);
        // round-trips to the same token count
        let total: usize = groups.iter().map(|g| g.len()).sum();
        assert_eq!(total, spans.len());
    }

    #[test]
    fn max_aggregator_drops_droppable_span() {
        let text = "keep this needle phrase but drop filler tokens here";
        let (ids, spans) = token_spans(text).unwrap();
        // uniform mid drop-prob, decode at 30% — should produce shorter text.
        let probs = vec![0.5_f64; ids.len()];
        let out = span_decode(
            &ids,
            &spans,
            text,
            &probs,
            0.3,
            "word",
            Aggregator::Max,
            None,
        )
        .unwrap();
        assert!(out.len() <= text.len());
    }

    #[test]
    fn force_keep_protects_span() {
        let text = "alpha beta gamma delta epsilon zeta";
        let (ids, spans) = token_spans(text).unwrap();
        let probs = vec![0.9_f64; ids.len()];
        let mut fk = vec![false; ids.len()];
        fk[0] = true; // lock first token
        let out = span_decode(
            &ids,
            &spans,
            text,
            &probs,
            1.0,
            "word",
            Aggregator::Max,
            Some(&fk),
        )
        .unwrap();
        assert!(out.contains("alpha"));
    }
}
